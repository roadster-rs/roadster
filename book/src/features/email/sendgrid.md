# Sendgrid

Sendgrid is a popular provider for sending email with a visual email template builder. When the `email-sendgrid` feature
is enabled, Roadster will initialize a Sendgrid client via the [`sendgrid`](https://docs.rs/sendgrid/0.23.0/sendgrid/)
crate.

Sendgrid also supports sending emails via SMTP. For details on Roadster's SMTP integration,
see [the previous chapter](./smtp.md).

## Configure the Sendgrid integration

The Sendgrid connection details can be configured via your app's config files, and via env vars or an [
`AsyncSource`](/features/configuration.html#custom-async-sources) for
sensitive connection details.

```toml
{{ #include ../../../examples/email/config/development/sendgrid.toml}}
```

## Sending an email

With the Sendgrid client, emails are sent by providing a Sendgrid email template ID and the template's parameters. Below
is a simple example of using the Sendgrid client.
See [Sendgrid's docs](https://www.twilio.com/docs/sendgrid/api-reference) for more details on the other fields you may
want to set when sending emails.

```rust,ignore
{{#include ../../../examples/email/src/worker/email/sendgrid.rs:12:}}
```

## Docs.rs links

- [email config](https://docs.rs/roadster/latest/roadster/config/email/struct.Email.html)
