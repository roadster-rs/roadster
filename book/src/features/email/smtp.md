# SMTP

Roadster's SMTP support allows sending emails with virtually any email provider. The caveat with using plain SMTP is
you may not be able to use the visual email builders provided by some vendors (e.g. Sendgrid or Customer.io). This
means that while you're code will be decoupled from any particular vendor, some additional work will be needed to send
"pretty" emails that match your app's design style. However, if you just need to send plain text emails, or are willing
to do the additional work to create "pretty" emails, SMTP is a great option for sending emails for your app.

## Starting a local SMTP service for testing

There are several SMTP servers that can be run locally for testing. This is another benefit of using SMTP instead of
Sendgrid -- a local SMTP instance can be used to easily verify the contents of your emails and your sending logic, while
Sendgrid only provides minimal dev/testing support and can't be run locally.

Below are a few options for development SMTP services that can be easily run locally with docker.

### [Mailpit](https://github.com/axllent/mailpit)

```shell
docker run -d -p 8025:8025 -p 1025:1025 axllent/mailpit
```

### [smtp4dev](https://github.com/rnwood/smtp4dev)

```shell
docker run -d -p 1080:80 -p 1025:25 rnwood/smtp4dev
```

### [maildev](https://github.com/maildev/maildev)

```shell
docker run -d -p 1080:1080 -p 1025:1025 maildev/maildev
```

## Configure the SMTP integration

The SMTP connection details can be configured via your app's config files, and via env vars or an [
`AsyncSource`](/features/configuration.html#custom-async-sources) for
sensitive connection details. Below is a sample config file that can be used to connect to a locally-hosted SMTP
service.

```toml
{{ #include ../../../examples/email/config/development/smtp.toml}}
```

## Sending plaintext emails

The easiest way to send an email is to simply send a plaintext email.

```rust,ignore
{{#include ../../../examples/email/src/worker/email/smtp.rs:15:}}
```

## Sending html emails with Leptos

In order to send HTML content via SMTP, you can either manually write your HTML, or use something
like [Leptos](https://docs.rs/leptos/latest/leptos/). Leptos is a reactive Rust UI framework, but it can also be used as
a simple HTML templating system. It may be possible to use other frameworks such
as [Yew](https://docs.rs/yew/latest/yew/) or [Dioxus](https://docs.rs/dioxus/latest/dioxus/) as well.

The below example is the same as the plaintext example, except it formats the email message with HTML using Leptos.

```rust,ignore
{{#include ../../../examples/email/src/worker/email/smtp_leptos.rs:16:}}
```

## Docs.rs links

- [email config](https://docs.rs/roadster/latest/roadster/config/email/struct.Email.html)