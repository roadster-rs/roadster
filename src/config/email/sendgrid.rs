use crate::util::serde::default_true;
use serde_derive::{Deserialize, Serialize};
use validator::Validate;

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct Sendgrid {
    pub api_key: String,
    #[serde(default = "default_true")]
    pub https_only: bool,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::snapshot::TestCase;
    use insta::assert_toml_snapshot;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(
        r#"
        api-key = "api-key"
        "#
    )]
    #[case(
        r#"
        api-key = "api-key"
        http_only = false
        "#
    )]
    #[case(
        r#"
        api-key = "api-key"
        http_only = true
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn serialization(_case: TestCase, #[case] config: &str) {
        let sendgrid: Sendgrid = toml::from_str(config).unwrap();

        assert_toml_snapshot!(sendgrid);
    }
}
