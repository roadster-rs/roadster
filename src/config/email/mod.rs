#[cfg(feature = "email-sendgrid")]
pub mod sendgrid;
#[cfg(feature = "email-smtp")]
pub mod smtp;

use lettre::message::Mailbox;
#[cfg(feature = "email-sendgrid")]
use sendgrid::Sendgrid;
use serde_derive::{Deserialize, Serialize};
#[cfg(feature = "email-smtp")]
use smtp::Smtp;
use validator::Validate;

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct Email {
    pub from: Mailbox,
    pub reply_to: Option<Mailbox>,
    #[cfg(feature = "email-smtp")]
    #[validate(nested)]
    pub smtp: Smtp,
    #[cfg(feature = "email-sendgrid")]
    #[validate(nested)]
    pub sendgrid: Sendgrid,
}

#[cfg(all(
    test,
    feature = "email",
    feature = "email-smtp",
    feature = "email-sendgrid"
))]
mod tests {
    use super::*;
    use crate::testing::snapshot::TestCase;
    use insta::assert_toml_snapshot;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(
        r#"
        [from]
        email = "no-reply@example.com"

        [smtp.connection]
        uri = "smtps://username:password@smtp.example.com:425"

        [sendgrid]
        api-key = "api-key"
        "#
    )]
    #[case(
        r#"
        from = "No Reply <no-reply@example.com>"

        [reply-to]
        email = "no-reply@example.com"
        name = "No Reply"

        [smtp.connection]
        uri = "smtps://username:password@smtp.example.com:425"

        [sendgrid]
        api-key = "api-key"
        "#
    )]
    #[case(
        r#"
        reply-to = "No Reply <no-reply@example.com>"

        [from]
        email = "no-reply@example.com"
        name = "No Reply"

        [smtp.connection]
        uri = "smtps://username:password@smtp.example.com:425"

        [sendgrid]
        api-key = "api-key"
        "#
    )]
    #[case(
        r#"
        from = "no-reply@example.com"

        [smtp.connection]
        uri = "smtps://username:password@smtp.example.com:425"

        [sendgrid]
        api-key = "api-key"
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn serialization(_case: TestCase, #[case] config: &str) {
        let email: Email = toml::from_str(config).unwrap();

        assert_toml_snapshot!(email);
    }
}
