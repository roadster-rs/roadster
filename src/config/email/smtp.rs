use crate::config::email::Email;
use lettre::message::MessageBuilder;
use lettre::transport::smtp::authentication::Credentials;
use lettre::transport::smtp::SmtpTransportBuilder;
use lettre::SmtpTransport;
use serde_derive::{Deserialize, Serialize};
use url::Url;
use validator::{Validate, ValidationErrors};

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct Smtp {
    #[validate(nested)]
    pub connection: SmtpConnection,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged, rename_all = "kebab-case")]
#[non_exhaustive]
pub enum SmtpConnection {
    Fields(SmtpConnectionFields),
    Uri(SmtpConnectionUri),
}

impl Validate for SmtpConnection {
    fn validate(&self) -> Result<(), ValidationErrors> {
        match self {
            SmtpConnection::Fields(fields) => fields.validate(),
            SmtpConnection::Uri(uri) => uri.validate(),
        }
    }
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct SmtpConnectionFields {
    pub host: String,
    pub port: Option<u16>,
    pub username: String,
    pub password: String,
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct SmtpConnectionUri {
    pub uri: Url,
}

impl From<&Email> for MessageBuilder {
    fn from(value: &Email) -> Self {
        let builder = MessageBuilder::new().from(value.from.clone());
        let builder = if let Some(reply_to) = value.reply_to.as_ref() {
            builder.reply_to(reply_to.clone())
        } else {
            builder
        };

        builder
    }
}

impl TryFrom<&SmtpConnection> for SmtpTransportBuilder {
    type Error = lettre::transport::smtp::Error;

    fn try_from(value: &SmtpConnection) -> Result<Self, Self::Error> {
        match value {
            SmtpConnection::Fields(fields) => {
                let credentials =
                    Credentials::new(fields.username.clone(), fields.password.clone());
                SmtpTransport::relay(&fields.host)
                    .map(|builder| {
                        if let Some(port) = fields.port {
                            builder.port(port)
                        } else {
                            builder
                        }
                    })
                    .map(|builder| builder.credentials(credentials))
            }
            SmtpConnection::Uri(fields) => SmtpTransport::from_url(fields.uri.as_ref()),
        }
    }
}

impl TryFrom<&SmtpConnection> for SmtpTransport {
    type Error = lettre::transport::smtp::Error;

    fn try_from(value: &SmtpConnection) -> Result<Self, Self::Error> {
        let builder: SmtpTransportBuilder = value.try_into()?;
        Ok(builder.build())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::snapshot::TestCase;
    use insta::assert_toml_snapshot;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(
        r#"
        [from]
        email = "no-reply@example.com"

        [smtp.connection]
        uri = "smtps://username:password@smtp.example.com:425"
        "#
    )]
    #[case(
        r#"
        from = "No Reply <no-reply@example.com>"

        [reply-to]
        email = "no-reply@example.com"
        name = "No Reply"

        [smtp.connection]
        host = "smtp.example.com"
        username = "username"
        password = "password"
        "#
    )]
    #[case(
        r#"
        [from]
        email = "no-reply@example.com"

        [smtp.connection]
        host = "smtp.example.com"
        username = "username"
        password = "password"
        "#
    )]
    #[case(
        r#"
        reply-to = "No Reply <no-reply@example.com>"

        [from]
        email = "no-reply@example.com"
        name = "No Reply"

        [smtp.connection]
        uri = "smtps://username:password@smtp.example.com:425"
        "#
    )]
    #[case(
        r#"
        from = "no-reply@example.com"

        [smtp.connection]
        uri = "smtps://username:password@smtp.example.com:425"
        "#
    )]
    #[case(
        r#"
        from = "no-reply@example.com"

        [smtp.connection]
        host = "smtp.example.com"
        port = 465
        username = "username"
        password = "password"
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn serialization(_case: TestCase, #[case] config: &str) {
        let email: Email = toml::from_str(config).unwrap();

        assert_toml_snapshot!(email);
    }
}
