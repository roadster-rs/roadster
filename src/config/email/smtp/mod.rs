use crate::config::email::Email;
use lettre::SmtpTransport;
use lettre::message::MessageBuilder;
use lettre::transport::smtp::authentication::Credentials;
use lettre::transport::smtp::{PoolConfig, SmtpTransportBuilder};
use serde_derive::{Deserialize, Serialize};
use serde_with::serde_as;
use std::time::Duration;
use url::Url;
use validator::{Validate, ValidationErrors};

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct Smtp {
    #[validate(nested)]
    pub connection: SmtpConnection,

    #[serde(default)]
    #[validate(nested)]
    pub pool: Option<SmtpPool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged, rename_all = "kebab-case")]
#[non_exhaustive]
pub enum SmtpConnection {
    Fields(SmtpConnectionFields),
    Uri(SmtpConnectionUri),
}

impl Validate for SmtpConnection {
    fn validate(&self) -> Result<(), ValidationErrors> {
        match self {
            SmtpConnection::Fields(fields) => fields.validate(),
            SmtpConnection::Uri(uri) => uri.validate(),
        }
    }
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct SmtpConnectionFields {
    pub host: String,
    pub port: Option<u16>,
    pub username: String,
    pub password: String,
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct SmtpConnectionUri {
    pub uri: Url,
}

#[serde_as]
#[derive(Debug, Default, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct SmtpPool {
    #[serde(default)]
    pub min_connections: Option<u32>,

    #[serde(default)]
    pub max_connections: Option<u32>,

    #[serde(default)]
    #[serde_as(as = "Option<serde_with::DurationSeconds>")]
    pub idle_timeout: Option<Duration>,
}

impl From<&Email> for MessageBuilder {
    fn from(value: &Email) -> Self {
        let builder = MessageBuilder::new().from(value.from.clone());
        let builder = if let Some(reply_to) = value.reply_to.as_ref() {
            builder.reply_to(reply_to.clone())
        } else {
            builder
        };

        builder
    }
}

impl TryFrom<&Smtp> for SmtpTransportBuilder {
    type Error = lettre::transport::smtp::Error;

    fn try_from(value: &Smtp) -> Result<Self, Self::Error> {
        let builder = match &value.connection {
            SmtpConnection::Fields(fields) => {
                let credentials =
                    Credentials::new(fields.username.clone(), fields.password.clone());
                SmtpTransport::relay(&fields.host)
                    .map(|builder| {
                        if let Some(port) = fields.port {
                            builder.port(port)
                        } else {
                            builder
                        }
                    })
                    .map(|builder| builder.credentials(credentials))
            }
            SmtpConnection::Uri(fields) => SmtpTransport::from_url(fields.uri.as_ref()),
        }?;

        let builder = if let Some(smtp_pool) = value.pool.as_ref() {
            builder.pool_config(smtp_pool.into())
        } else {
            builder
        };

        Ok(builder)
    }
}

impl From<&SmtpPool> for PoolConfig {
    fn from(value: &SmtpPool) -> Self {
        let pool_config = PoolConfig::new();

        let pool_config = if let Some(min_connections) = value.min_connections {
            pool_config.min_idle(min_connections)
        } else {
            pool_config
        };

        let pool_config = if let Some(max_connections) = value.max_connections {
            pool_config.max_size(max_connections)
        } else {
            pool_config
        };

        if let Some(idle_timeout) = value.idle_timeout {
            pool_config.idle_timeout(idle_timeout)
        } else {
            pool_config
        }
    }
}

impl TryFrom<&Smtp> for SmtpTransport {
    type Error = lettre::transport::smtp::Error;

    fn try_from(value: &Smtp) -> Result<Self, Self::Error> {
        let builder: SmtpTransportBuilder = value.try_into()?;
        Ok(builder.build())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::snapshot::TestCase;
    use insta::{assert_debug_snapshot, assert_toml_snapshot};
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(
        r#"
        [connection]
        uri = "smtps://username:password@smtp.example.com:425"
        "#
    )]
    #[case(
        r#"
        [connection]
        host = "smtp.example.com"
        username = "username"
        password = "password"
        "#
    )]
    #[case(
        r#"
        [connection]
        host = "smtp.example.com"
        port = 465
        username = "username"
        password = "password"
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn serialization(_case: TestCase, #[case] config: &str) {
        let smtp: Smtp = toml::from_str(config).unwrap();
        smtp.validate().unwrap();

        assert_toml_snapshot!(smtp);
    }

    #[rstest]
    #[case(
        r#"
        from = "foo@example.com"
        [smtp.connection]
        uri = "smtps://username:password@smtp.example.com:425"
        [sendgrid]
        api-key = "api-key"
        "#
    )]
    #[case(
        r#"
        from = "foo@example.com"
        reply-to = "no-reply@example.com"
        [smtp.connection]
        uri = "smtps://username:password@smtp.example.com:425"
        [sendgrid]
        api-key = "api-key"
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn message_builder_from_email(_case: TestCase, #[case] email_config: &str) {
        let email: Email = toml::from_str(email_config).unwrap();
        let msg_builder: MessageBuilder = (&email).into();
        assert_debug_snapshot!(msg_builder);
    }

    #[rstest]
    #[case(
        r#"
        [connection]
        uri = "smtps://username:password@smtp.example.com:425"
        "#
    )]
    #[case(
        r#"
        [connection]
        host = "smtp.example.com"
        port = 465
        username = "username"
        password = "password"
        "#
    )]
    #[case(
        r#"
        [connection]
        host = "smtp.example.com"
        username = "username"
        password = "password"
        "#
    )]
    #[case(
        r#"
        [connection]
        uri = "smtps://username:password@smtp.example.com:425"
        [pool]
        min-connections = 1
        "#
    )]
    #[case(
        r#"
        [connection]
        uri = "smtps://username:password@smtp.example.com:425"
        [pool]
        max-connections = 100
        "#
    )]
    #[case(
        r#"
        [connection]
        uri = "smtps://username:password@smtp.example.com:425"
        [pool]
        idle-timeout = 60
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn smtp_connection_from_config(_case: TestCase, #[case] smtp_config: &str) {
        let smtp: Smtp = toml::from_str(smtp_config).unwrap();
        let _smtp_transport: SmtpTransport = (&smtp).try_into().unwrap();
    }

    #[rstest]
    #[case(
        r#"
        [connection]
        uri = "https://username:password@smtp.example.com:425"
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn smtp_connection_from_config_err(_case: TestCase, #[case] smtp_config: &str) {
        let smtp: Smtp = toml::from_str(smtp_config).unwrap();
        assert!(SmtpTransport::try_from(&smtp).is_err());
    }
}
