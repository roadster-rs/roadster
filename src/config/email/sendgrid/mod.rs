use crate::config::environment::Environment;
use crate::util::serde::default_true;
use config::{FileFormat, FileSourceString};
use serde_derive::{Deserialize, Serialize};
use validator::Validate;

pub(crate) fn default_config_per_env(
    environment: Environment,
) -> Option<config::File<FileSourceString, FileFormat>> {
    let config = match environment {
        Environment::Production => Some(include_str!("config/production.toml")),
        _ => None,
    };
    config.map(|c| config::File::from_str(c, FileFormat::Toml))
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[non_exhaustive]
pub struct Sendgrid {
    pub api_key: String,
    #[serde(default = "default_true")]
    pub sandbox: bool,
    #[serde(default = "default_true")]
    pub https_only: bool,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::snapshot::TestCase;
    use insta::assert_toml_snapshot;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case(
        r#"
        api-key = "api-key"
        "#
    )]
    #[case(
        r#"
        api-key = "api-key"
        http_only = false
        "#
    )]
    #[case(
        r#"
        api-key = "api-key"
        http_only = true
        "#
    )]
    #[case(
        r#"
        api-key = "api-key"
        sandbox = false
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn serialization(_case: TestCase, #[case] config: &str) {
        let sendgrid: Sendgrid = toml::from_str(config).unwrap();

        assert_toml_snapshot!(sendgrid);
    }
}
