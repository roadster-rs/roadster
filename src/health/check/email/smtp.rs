use crate::api::core::health::smtp_health;
use crate::app::context::{AppContext, AppContextWeak};
use crate::error::RoadsterResult;
use crate::health::check::{missing_context_response, CheckResponse, HealthCheck};
use async_trait::async_trait;
use tracing::instrument;

pub struct SmtpHealthCheck {
    pub(crate) context: AppContextWeak,
}

#[async_trait]
impl HealthCheck for SmtpHealthCheck {
    fn name(&self) -> String {
        "smtp".to_string()
    }

    fn enabled(&self) -> bool {
        self.context
            .upgrade()
            .map(|context| enabled(&context))
            .unwrap_or_default()
    }

    #[instrument(skip_all)]
    async fn check(&self) -> RoadsterResult<CheckResponse> {
        let context = self.context.upgrade();
        let response = match context {
            Some(context) => smtp_health(&context, None).await,
            None => missing_context_response(),
        };
        Ok(response)
    }
}

fn enabled(context: &AppContext) -> bool {
    context.config().health_check.smtp.common.enabled(context)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::AppConfig;
    use rstest::rstest;

    #[rstest]
    #[case(false, Some(true), true)]
    #[case(false, Some(false), false)]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn enabled(
        #[case] default_enable: bool,
        #[case] enable: Option<bool>,
        #[case] expected_enabled: bool,
    ) {
        // Arrange
        let mut config = AppConfig::test(None).unwrap();
        config.health_check.default_enable = default_enable;
        config.health_check.smtp.common.enable = enable;

        let context = AppContext::test(Some(config), None, None).unwrap();

        // Act/Assert
        assert_eq!(super::enabled(&context), expected_enabled);
    }
}
