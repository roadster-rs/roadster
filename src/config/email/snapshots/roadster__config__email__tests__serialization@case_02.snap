---
source: src/config/email/mod.rs
expression: email
---
from = 'No Reply <no-reply@example.com>'
reply-to = 'No Reply <no-reply@example.com>'
[smtp.connection]
uri = 'smtps://username:password@smtp.example.com:425'

[sendgrid]
api-key = 'api-key'
sandbox = true
https-only = true
